### PR TITLE
allow records to have type annotations

### PR DESCRIPTION
**Pull request by 1Kinoti**
# Description
follow up to #8529
cleaned up version of #8892 

- the original syntax is okay
```nu
def okay [rec: record] {}
```
- you can now add type annotations for fields if you know
  them before hand
```nu
def okay [rec: record<name: string>] {}
```

- you can specify multiple fields
```nu
def okay [person: record<name: string age: int>] {}

# an optional comma is allowed
def okay [person: record<name: string, age: int>] {}
```

- if annotations are specified, any use of the command will be type
  checked against the specified type
```nu
def unwrap [result: record<ok: bool, value: any>] {}

unwrap {ok: 2, value: "value"}

# errors with

Error: nu::parser::type_mismatch

  × Type mismatch.
   ╭─[entry #4:1:1]
 1 │ unwrap {ok: 2, value: "value"}
   ·         ───────┬─────
   ·                    ╰── expected record<ok: bool, value: any>, found record<ok: int, value: string>
   ╰────
```
> here the error is in the `ok` field, since `any` is coerced into any type
> as a result `unwrap {ok: true, value: "value"}` is okay

- the key must be a string, either quoted or unquoted
```nu
def err [rec: record<{}: list>] {}

# errors with
Error:
  × `record` type annotations key not string
   ╭─[entry #7:1:1]
 1 │ def unwrap [result: record<{}: bool, value: any>] {}
   ·                            ─┬
   ·                             ╰── must be a string
   ╰────
```

- a key doesn't have to have a type in which case it is assumed to be `any`
```nu
def okay [person: record<name age>] {}

def okay [person: record<name: string age>] {}
```

- however, if you put a colon, you have to specify a type
```nu
def err [person: record<name: >] {}

# errors with
Error: nu::parser::parse_mismatch

  × Parse mismatch during operation.
   ╭─[entry #12:1:1]
 1 │ def unwrap [res: record<name: >] { $res }
   ·                             ┬
   ·                             ╰── expected type after colon
   ╰────
```

# User-Facing Changes
**[BREAKING CHANGES]**
- this change adds a field to `SyntaxShape::Record` so any plugins that used it will have to update and include the field. though if you are unsure of the type the record expects, `SyntaxShape::Record(vec![])` will suffice

### CODE REVIEW

**Comment by jntrnr** [comment]
We were talking about this in a recent design meeting, and we decided to move away from the `<T>` generic style. Instead, we'd list off the types in a name:value style:

```
table(name: string, age: int)
```

or
```
record(name: string, id: string)
```

or
```
list(int)
```

We talked a bit about making the last one `[int]` but I think decided to go with the readable names for now.

File: src/tests/test_signatures.rs [hunk 1]
```diff
@@ -23,112 +23,298 @@ fn list_annotations_empty_1() -> TestResult {
...
+fn record_annotations_none() -> TestResult {
+    let input = "def run [rec: record] { $rec }; run {} | describe";
+    let expected = "record";
+    run_test(input, expected)
+}
+
+#[test]
+fn record_annotations() -> TestResult {
+    let input = "def run [rec: record(age: int)] { $rec }; run {age: 3} | describe";
+    let expected = "record<age: int>";
```
Comment (jntrnr) [inline src/tests/test_signatures.rs:153]: these error messages should also use the `record(age: int)` syntax.

**Comment by jntrnr** [comment]
Nice, I'm glad it cleans up some code, too.

**Comment by bobhy** [comment]
Wrapping these shape decls for `table` and `record` in parens seems odd compared to  `def <name> [ <arglist> ]`.  Why should these delimiters be different?
But rather than change the parens to square brackets on shape decls, I'd go the other way and use *parens* in `def` arg decls.  I've always thought the square brackets in this context were non-industry-standard. I get that it's evocative of a *list*, but still...  
Is it just too late to change?

**Comment by amtoine** [comment]
can't wait to see these new annotations in `nushell` @1Kinoti :star_struck:

**Comment by jntrnr** [comment]
> Wrapping these shape decls for `table` and `record` in parens seems odd compared to `def <name> [ <arglist> ]`. Why should these delimiters be different? But rather than change the parens to square brackets on shape decls, I'd go the other way and use _parens_ in `def` arg decls. I've always thought the square brackets in this context were non-industry-standard. I get that it's evocative of a _list_, but still... Is it just too late to change?

Not sure what you're asking, but if you are asking about reusing delimiters, you wouldn't want the same delimiter for both the argument list and the type parameters to the types.  Readability improves by using two separate ones`[foo: Record[name: string, id: int]]` is much harder to read than `[foo: Record(name: string, id: int)]`

**Comment by bobhy** [comment]
Let me take the broader question up on Discord, it's not a problem with this PR specifically.   It's great for Nu to be able to annotate record and table shapes!

**Comment by fdncred** [comment]
I remember having this discussion about this change but what I don't remember is why we changed from `<>` to `()`. Seems like `list<int>` is more intuitive than `list(int)` since the parentheses makes it looks like `list()` is a function with the `int` parameter.

Also, since we allow optional parentheses in custom command definition, like:
```
def ok (rec: record(name: string)) { $rec }
```
i'm not sure that's more readable than this below. It kind of feels the same.
```
def ok [rec: record[name: string]] { $rec }
```

But
```
def ok (rec: record<name: string>) { $rec }
```
and 
```
def ok [rec: record<name: string>] { $rec }
```
both still read as `rec` is the parameter with a type of `record<name: string>`.

I'm not willing to die on this hill, but I think the "alligators" <> are more intuitive than the parentheses, in this use case.

**Comment by bobhy** [comment]
I think the reason we're going back and forth on this is that we lack an agreed overarching model for the language. 
 Do we want NU to "just look right" as a thing unto itself?  or should it "remind users of" some other language, and, if so, which (language or languages)?  
For now, I say, ship it!  We have a *working* feature (modulo the final syntax), and the strong possibility that resolution of the syntax question will require more wide-spread changes when it comes.  Might as well get the benefit of the (stronger typing) behavior in the meantime.

**Comment by jntrnr** [comment]
Just to 👍 what @fdncred said - after we talked about it in a recent design meeting, we realised that that `()` for the type parameter would look too confusing if folks used the alternate parameter syntax `def foo (x: list(int)) { ... }`

We've been talking about also adding an alternative call syntax like `foo([1, 2, 3])` so that might encourage more folks to use the `()` syntax.

As a result, seems like `<>` is our best bet for the parameter syntax for now.

Sorry about going back and forth.

**Comment by amtoine** [comment]
and using `<>` back might have the side-effect of making this PR not a `breaking-change` :+1:

**Comment by 1Kinoti** [comment]
> Just to +1 what @fdncred said - after we talked about it in a recent design meeting, we realised that that () for the type parameter would look too confusing if folks used the alternate parameter syntax def foo (x: list(int)) { ... }
> 
> We've been talking about also adding an alternative call syntax like foo([1, 2, 3]) so that might encourage more folks to use the () syntax.
> 
> As a result, seems like <> is our best bet for the type parameter syntax for now.
> 
> Sorry about going back and forth.

no problem @jntrnr , let me do that

**Comment by fdncred** [comment]
Thanks @1Kinoti. We're close to a release (April 25), and we have a 2 day freeze before that day, so we probably won't land this until after that. I'd like to have more than a day to test it. Thanks again for all your support and PRs!!!

**Comment by 1Kinoti** [comment]
> Thanks @1Kinoti. We're close to a release (April 25), and we have a 2 day freeze before that day, so we probably won't land this until after that. I'd like to have more than a day to test it. Thanks again for all your support and PRs!!!

you're welcome @fdncred! after the release it is

**Comment by fdncred** [comment]
@1Kinoti Thanks again. I'm loving this change!

**Comment by amtoine** [comment]
can't wait to play with them :star_struck:

**Pull request merged**